CNV-39643: Document generlizing a Linux VM (#81924)
openshift/openshift-docs: modules/virt-generalizing-linux-vm-image.adoc

// Module included in the following assemblies:
//
// * virt/virtual_machines/creating_vms_custom/virt-creating-vms-uploading-images.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-generalizing-linux-vm-image_{context}"]
= Generalizing a VM image

You can generalize a {op-system-base-full} image to remove all system-specific configuration data before you use the image to create a golden image, a preconfigured snapshot of a virtual machine (VM). You can use a golden image to deploy new VMs.

> Generalizing a {op-system-base} VM is done using the `virtctl`, `guestfs`, and `sys-prep` tools.

Reviewer: ```suggestion
Generalizing a {op-system-base} VM is done using the `virtctl`, `guestfs`, and `virt-sysprep` tools.
```

.Prerequisites

> * A running {op-system-base} virtual machine.

Reviewer: ```suggestion
* A {op-system-base} virtual machine.
```
running is not required

* You have installed the OpenShift CLI (`oc`).
* You have installed the `virtctl` tool.

.Procedure

. Stop the {op-system-base} virtual machine, if it is not already stopped.
+
[source,terminal]
----
$ virtctl stop <rhel-vm-name>
----

> . Optional: Clone the virtual machine to avoid losing the data from your original VM. You can then generalize the cloned VM.

Reviewer 2: since we don't share cloning procedure maybe hyperlink to already existing cloning procedure.

. Retrieve the Persistent Volume Claim (PVC) that stores the root filesystem for the VM.
+
[source,terminal]
----
$ oc get vm <my-vm-name> -o jsonpath="{.spec.template.spec.volumes}{'\n'}"

[{"dataVolume":{"name":"my-vm-volume"},"name":"rootdisk"},{"cloudInitNoCloud":{...}]
----

. Retrieve the `dataVolume` that matches the listed PVCs.
+
[source,terminal]
----
$ oc get pvc
NAME            STATUS   VOLUME  CAPACITY   ACCESS MODES  STORAGECLASS     AGE
my-vm-volume Bound  …
----

[NOTE]
====
If your cluster configuration does not enable you to clone a VM, and you want to avoid losing the data from your original VM, you can clone the VM PVC to a data volume instead. You can then use the cloned PVC to create a golden image.

If you are creating a golden image by cloning a PVC, continue with the steps below, using the cloned PVC.
====

> . Deploy an interactive container with the PVC and `ibguestfs-tools` attached to it.

Reviewer 2: is it libguestfs-tools?

Reviewer 2: Also do we plan to provide hyperlinks like we have in document?

Author: We can't include hyperlinks in the module itself. I have a section added at the end of this procedure with relevant links to cloning a VM and cloning a PVC. I can show you a preview of what it looks like once we work out the issues eith my failing prow checks on this PR.

Author: @dominikholler can you comment regarding libguestfs-tools?

Reviewer: ```suggestion
. Deploy an interactive container with the PVC and `libguestfs-tools` attached to it.
```
I guess @geetikakay is referring to the missing "l"

Reviewer 2: yes

+
[source,terminal]
----

> $ virtctl guestfs my-vm-volume –uid 107

Reviewer 2: --uid

----

. Remove all configurations specific to your system.

> +

Reviewer: This is correct, I just wonder if it could  be highlighted that the ` virt-sysprep -a disk.img` should be executed in the session/shell opened by the previous command.

// Module included in the following assemblies:
//
// * virt/virtual_machines/creating_vms_custom/virt-creating-vms-uploading-images.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-generalizing-linux-vm-image_{context}"]
= Generalizing a VM image

You can generalize a {op-system-base-full} image to remove all system-specific configuration data before you use the image to create a golden image, a preconfigured snapshot of a virtual machine (VM). You can use a golden image to deploy new VMs.

Generalizing a {op-system-base} VM is done using the `virtctl`, `guestfs`, and `virt-sysprep` tools.

.Prerequisites

> * An {op-system-base} virtual machine.

Author: Since {op-system--base] renders as Openshift container platform, is actually should be "An" not "A".

> * An {op-system-base} virtual machine.

Reviewer 3: I must be missing something -- It's rendering as "An RHEL virtual machine."

> * An {op-system-base} virtual machine.

Author: Oh, my mistake then, you're right :woman_facepalming:

> * An {op-system-base} virtual machine.

Author: This PR has been open so long I can barely see straight anymore :sweat_smile:

> * An {op-system-base} virtual machine.

Author: I did update it to read "You have a {op-system-base} virtual machine (VM) to use as a base VM."

* You have installed the OpenShift CLI (`oc`).
* You have installed the `virtctl` tool.

.Procedure

> . Stop the {op-system-base} virtual machine, if it is not already stopped.

Reviewer 4: Give the "why" before the "what". So rewrite as: "If the {op-system-base} virtual machine is running, stop it."

+
[source,terminal]
----
$ virtctl stop <rhel-vm-name>
----

> . Optional: Clone the virtual machine to avoid losing the data from your original VM. You can then generalize the cloned VM.

Author: I'm going to leave this as is, since I think the alternation of virtual machine and VM (since they're mentioned 3 times) actually makes this easier for the user to parse.

> . Retrieve the `dataVolume` that stores the root filesystem for the VM.

Reviewer 4: Add "...by running the following command:"

> . Retrieve the Persistent Volume Claim (PVC) that stores the root filesystem for the VM.

Reviewer 5: ```suggestion
. Retrieve the `dataVolume` that stores the root filesystem for the VM.
```

+
[source,terminal]
----
$ oc get vm <my-vm-name> -o jsonpath="{.spec.template.spec.volumes}{'\n'}"

> [{"dataVolume":{"name":"my-vm-volume"},"name":"rootdisk"},{"cloudInitNoCloud":{...}]

Reviewer 3: This is the example output, correct? 

Please use the `.Example output` heading and a separate code block for it.

Author: @dominikholler Is this example output or part of the command?

Reviewer: This is example output.

----

> . Retrieve the `dataVolume` that matches the listed PVCs.

Reviewer 5: ```suggestion
. Retrieve the Persistent Volume Claim (PVC) that matches the listed `dataVolume`.
```

> . Retrieve the Persistent Volume Claim (PVC) that matches the listed `dataVolume`.

Reviewer 4: Same comment as above.

+
[source,terminal]
----
$ oc get pvc
NAME            STATUS   VOLUME  CAPACITY   ACCESS MODES  STORAGECLASS     AGE
my-vm-volume Bound  …
----

> [NOTE]

Reviewer 3: <img width="703" alt="Screenshot 2024-10-08 at 3 02 48 PM" src="https://github.com/user-attachments/assets/287d5f36-7c3f-4c94-ad57-e02c9e2a1403">

This note is not rendering as a part of the Step.  I think using a + will make it aligned to the step.

For example like this - 

<img width="703" alt="Screenshot 2024-10-08 at 3 06 19 PM" src="https://github.com/user-attachments/assets/c406ea32-bf58-422f-8802-b188ea30236c">

Author: It's not meant to be part of the step, actually. It's not related to the step before it, but the steps after the note are necessary if the user is cloning a PVC instead of a VM. This was a tricky bit of documentation to make clear, and this was the best option. The devs were very happy with it.

Reviewer 3: The note looks out of place. I don't think this pattern of rendering is allowed. However you can check on the channel if this note can be kept or need to be moved to another place...

Author: I'll indent it. I don't think it will matter for the end user.

====

> If your cluster configuration does not enable you to clone a VM, and you want to avoid losing the data from your original VM, you can clone the VM PVC to a data volume instead. You can then use the cloned PVC to create a golden image.

Reviewer 3: ```suggestion
If your cluster configuration does not enable you to clone a VM, to avoid losing the data from your original VM, you can clone the VM PVC to a data volume. You can then use the cloned PVC to create a golden image.
```

Author: This is stylistic. I did remove the "and you want", but I'd rather leave in the "instead" for clarity.

> If you are creating a golden image by cloning a PVC, continue with the steps below, using the cloned PVC.

Reviewer 3: ```suggestion
If you are creating a golden image by cloning a PVC, continue with the next steps.
```
- I would avoid using "below". Consider using the "next steps" or "following steps".
- "using a cloned PVC" is redundant. "...creating a golden image **by cloning a PVC**" already gives context.

Author: I don't feel that "using the cloned PVC" is necessarily redundant, since the user could mistake needing to use the original PVC in the next steps. I'd prefer to keep that bit for clarity as well. I did change "below".

====

[start=5]

> . Deploy an interactive container with the PVC and `libguestfs-tools` attached to it.

Reviewer 4: Same comment as above.

> . Deploy an interactive container with the PVC and `libguestfs-tools` attached to it by running the following command:

Reviewer 3: ```suggestion
. Deploy an interactive container by attaching the PVC and `libguestfs-tools` to it. Run the following command:
```
IMO "...attached to it" might mean that pvc and guestfs are already attahed to it and user is just deploying the container.
Hence suggested "by attaching the PVC and ...."

Author: @dominikholler can you confirm?

Reviewer: There is a new interactive container from the `libguestfs-tools` created, and the PVC attached to it.

Author: So would this be correct?

**Deploy a new interactive container with `libguestfs-tools` and attach the PVC to it by running the following command:**

Reviewer: sounds good, thanks!

+
[source,terminal]
----

> $ virtctl guestfs my-vm-volume --uid 107

Reviewer 3: ```suggestion
$ virtctl guestfs <my_pvc_name> --uid 107
```
I think this should be a variable for the clarity and consistency.

Author: PVC and `dataVolume` are 2 different variables here. I made "my_vm_volume" a variable in earlier steps and updated the hyphens to underscores.

Author: I'll have the SME review this again after I merge the update to make sure.

> ----
> 

Reviewer 3: I guess this one is up to you. I thought of giving an idea about the opening shell beforehand. And in the next step, it would be clear that the command is to be run in the shell.

```suggestion
----
+
This command opens a shell for you to run the next command.
 
```

Author: I actually like this suggestion a lot, thanks.

> . In the shell opened using the previous command, remove all configurations specific to your system.

Reviewer 4: Same comment

> . In the shell opened using the previous command, remove all configurations specific to your system by running the following command:

Reviewer 3: ```suggestion
. Remove all configurations specific to your system by running the following command:
```

. Retrieve the `dataVolume` that stores the root filesystem for the VM by running the following command:
+
[source,terminal]
----
$ oc get vm <my_vm_name> -o jsonpath="{.spec.template.spec.volumes}{'\n'}"

[{"dataVolume":{"name":"<my_vm_volume>"},"name":"rootdisk"},{"cloudInitNoCloud":{...}]
----

. Retrieve the persistent volume claim (PVC) that matches the listed `dataVolume` by running the followimg command:
+
[source,terminal]
----
$ oc get pvc
----
+
.Example output
[source,terminal]
----
NAME            STATUS   VOLUME  CAPACITY   ACCESS MODES  STORAGECLASS     AGE
<my_vm_volume> Bound  …
----

[NOTE]
====
If your cluster configuration does not enable you to clone a VM, to avoid losing the data from your original VM, you can clone the VM PVC to a data volume instead. You can then use the cloned PVC to create a golden image.

If you are creating a golden image by cloning a PVC, continue with the next steps, using the cloned PVC.
====

[start=5]
. Deploy an interactive container with the PVC and `libguestfs-tools` attached to it by running the following command:
+
[source,terminal]
----

> $ virtctl guestfs <my-vm-volume> --uid 107

Reviewer: this is the my_vm_volume which was shown in the output from ` oc get pvc`

> $ virtctl guestfs <my-vm-volume> --uid 107

Author: Is that also the my_vm_volume from 
`oc get vm <my_vm_name> -o jsonpath="{.spec.template.spec.volumes}{'\n'}"`
?

> $ virtctl guestfs <my-vm-volume> --uid 107

Reviewer: @jcanocan can you help me here answering this question?

> $ virtctl guestfs <my-vm-volume> --uid 107

Reviewer 5: The command: 

`oc get vm <my_vm_name> -o jsonpath="{.spec.template.spec.volumes}{'\n'}"`

gives you the `DataVolume` used by the VM to storage the filesystem, and the command:

`oc get pvc`

gives you the underlying PVC used by that `DataVolume`.

Normally, both outputs should be the same. The reason to ask the user to run both commands is for safety reasons (it might include an unexpected "_", "-", etc. symbol). The user should use the matching output from `oc get pvcs` command.

Sorry about the confusion. My bad :(
